incuna_auth/middleware/permission.py

```diff
@@ -0,0 +1,215 @@
+from django.contrib import messages
+from django.http import HttpResponseForbidden, HttpResponseRedirect
+from django.utils.translation import ugettext_lazy as _
+
+from .utils import compile_urls
+from ..models import AccessStateExtensionMixin as AccessState
+
+
+ALL_URLS = compile_urls([r'^'])
+
+
+class BasePermissionMiddleware:
+    """
+    Base class for middleware that allows or denies access to a resource.
+
+    This is a generic, fairly useless base class. It protects no pages and allows all
+    users through, although if by some miracle it does deny access to a user, it'll be
+    fine doing that.
+
+    Contains the following implemented methods:
+    - process_request: the method called on incoming request.
+    - deny_access: provides standard "you're not allowed" responses.
+
+    Contains the following hook methods:
+    - is_resource_protected: Returns True if and only if the middleware should be
+      protecting the resource the request is trying to access.
+    - deny_access_condition: Returns True if and only if the request should be disallowed
+      from accessing the protected resource.
+    - get_unauthorised_redirect_url: Returns the URL to redirect a denied GET request to
+      (default implementation returns /).
+    - get_access_denied_message: Returns the message to display when that happens
+      (default implementation returns '').
+    """
+    def is_resource_protected(self, request, **kwargs):
+        """
+        Hook. Returns True if and only if the middleware should be protecting
+        the page or endpoint the request is trying to access.
+        """
+        return False
+
+    def deny_access_condition(self, request, **kwargs):
+        """
+        Hook. Returns True if and only if the request should be disallowed
+        from accessing the protected URL/page/endpoint.
+        """
+        return False
+
+    def get_unauthorised_redirect_url(self):
+        return '/'
```

Comment:
That said, I think it may need to be passed through to a "resolve" method, or something. Seem to remember this being fixed recently... was it @KevinEtienne that did it?